tycho-gpg-plugin/src/main/java/org/eclipse/tycho/gpg/SignRepositoryArtifactsMojo.java

```diff
@@ -30,213 +26,269 @@
+
+            var compressed = "true".equals(artifactRepository.getProperty(IRepository.PROP_COMPRESSED));
+
+            if (addPublicKeyToRepo && this.pgpKeyBehavior != PGPKeyBehavior.replace) {
+                var repositoryKeys = artifactRepository.getProperty(PGPSignatureVerifier.PGP_SIGNER_KEYS_PROPERTY_NAME);
+                if (repositoryKeys != null) {
+                    keys.add(repositoryKeys);
+                }
             }
-        }
-        Xpp3Dom dom = null;
-        try (var stream = new FileInputStream(artifactsXml)) {
-            dom = Xpp3DomBuilder.build(stream, StandardCharsets.UTF_8.displayName());
-        } catch (IOException | XmlPullParserException e) {
-            throw new MojoExecutionException(e.getMessage(), e);
-        }
-        ProxySignerWithPublicKeyAccess signer = newSigner(project);
-        String armoredPublicKey = signer.getPublicKeys();
-        for (var artifact : dom.getChild("artifacts").getChildren("artifact")) {
-            Xpp3Dom properties = artifact.getChild("properties");
-            if (Arrays.stream(properties.getChildren())
-                    .anyMatch(property -> IArtifactDescriptor.FORMAT.equals(property.getAttribute("name"))
-                            && IArtifactDescriptor.FORMAT_PACKED.equals(property.getAttribute("value")))) {
-                // skip packed artifacts
-                continue;
+
+            var artifactKeys = artifactRepository.query(ArtifactKeyQuery.ALL_KEYS, null);
+            var descriptors = StreamSupport.stream(artifactKeys.spliterator(), false)
+                    .map(artifactRepository::getArtifactDescriptors).map(Arrays::asList).flatMap(Collection::stream)
+                    .collect(Collectors.toList());
+            descriptors.parallelStream().forEach(
+                    it -> handle(it, artifactRepository.getArtifactFile(it), signedContentFactory, signer, keys));
+
+            if (addPublicKeyToRepo && !keys.isEmpty()) {
+                artifactRepository.setProperty(PGPSignatureVerifier.PGP_SIGNER_KEYS_PROPERTY_NAME,
+                        keys.toArmoredString());
             }
-            /*
-             * Different types of artifact have different locations in the repo. So we need to check
-             * the classifier to correctly get the location of the jar file.
-             * 
-             * Could potentially do this dynamically based on the mappings attribute with the repo.
-             */
-            String subDir = null;
-            switch (artifact.getAttribute("classifier")) {
-            case "osgi.bundle":
-                subDir = "plugins";
-                break;
-            case "org.eclipse.update.feature": //Not yet signing features
-            case "binary": //Not yet signing binaries
-            default:
-                continue; // Skip signing
+
+            artifactRepository.getClass().getMethod("save").invoke(artifactRepository);
```

Comment:
Can you open a issue at P2 so we probably get a way to invoke save without reflection?

Reply:
The problem here is that if I cast to something with a save method, I get an error about class loader problems. There is already a public save method.

Reply:
Can you share the error you get? It might be that we need to export a package from the `tycho-maven-plugin` that is `tycho-maven-plugin/src/main/resources/META-INF/maven/extension.xml` the maven way to share classes is really ugly :-\

Reply:
If I change it to this:

![image](https://user-images.githubusercontent.com/208716/203725790-b2c10d29-43e7-4bee-9726-70d3aae9f220.png)

The following is what happens, with the key message being `org.eclipse.equinox.internal.p2.artifact.repository.simple.SimpleArtifactRepository cannot be cast to class org.eclipse.equinox.internal.p2.artifact.repository.simple.SimpleArtifactRepository (org.eclipse.equinox.internal.p2.artifact.repository.simple.SimpleArtifactRepository is in unnamed module of loader org.codehaus.plexus.classworlds.realm.ClassRealm @7b451bf4; org.eclipse.equinox.internal.p2.artifact.repository.simple.SimpleArtifactRepository is in unnamed module of loader org.codehaus.plexus.classworlds.realm.ClassRealm @4987383f)`:
```
08:57:53 [ERROR] Failed to execute goal org.eclipse.tycho:tycho-gpg-plugin:4.0.0-SNAPSHOT:sign-p2-artifacts (default) on project org.eclipse.tycho.maven.site: A type incompatibility occurred while executing org.eclipse.tycho:tycho-gpg-plugin:4.0.0-SNAPSHOT:sign-p2-artifacts: class org.eclipse.equinox.internal.p2.artifact.repository.simple.SimpleArtifactRepository cannot be cast to class org.eclipse.equinox.internal.p2.artifact.repository.simple.SimpleArtifactRepository (org.eclipse.equinox.internal.p2.artifact.repository.simple.SimpleArtifactRepository is in unnamed module of loader org.codehaus.plexus.classworlds.realm.ClassRealm @7b451bf4; org.eclipse.equinox.internal.p2.artifact.repository.simple.SimpleArtifactRepository is in unnamed module of loader org.codehaus.plexus.classworlds.realm.ClassRealm @4987383f)
08:57:53 [ERROR] -----------------------------------------------------
08:57:53 [ERROR] realm =    plugin>org.eclipse.tycho:tycho-gpg-plugin:4.0.0-SNAPSHOT
08:57:53 [ERROR] strategy = org.codehaus.plexus.classworlds.strategy.SelfFirstStrategy
08:57:53 [ERROR] urls[0] = file:/D:/Users/merks/.m2/repository/org/eclipse/tycho/tycho-gpg-plugin/4.0.0-SNAPSHOT/tycho-gpg-plugin-4.0.0-SNAPSHOT.jar
08:57:53 [ERROR] urls[1] = file:/D:/Users/merks/.m2/repository/org/codehaus/plexus/plexus-utils/3.5.0/plexus-utils-3.5.0.jar
08:57:53 [ERROR] urls[2] = file:/D:/Users/merks/.m2/repository/org/codehaus/plexus/plexus-archiver/4.6.0/plexus-archiver-4.6.0.jar
08:57:53 [ERROR] urls[3] = file:/D:/Users/merks/.m2/repository/org/codehaus/plexus/plexus-io/3.4.0/plexus-io-3.4.0.jar
08:57:53 [ERROR] urls[4] = file:/D:/Users/merks/.m2/repository/commons-io/commons-io/2.11.0/commons-io-2.11.0.jar
08:57:53 [ERROR] urls[5] = file:/D:/Users/merks/.m2/repository/org/apache/commons/commons-compress/1.22/commons-compress-1.22.jar
08:57:53 [ERROR] urls[6] = file:/D:/Users/merks/.m2/repository/org/iq80/snappy/snappy/0.4/snappy-0.4.jar
08:57:53 [ERROR] urls[7] = file:/D:/Users/merks/.m2/repository/org/tukaani/xz/1.9/xz-1.9.jar
08:57:53 [ERROR] urls[8] = file:/D:/Users/merks/.m2/repository/com/github/luben/zstd-jni/1.5.2-5/zstd-jni-1.5.2-5.jar
08:57:53 [ERROR] urls[9] = file:/D:/Users/merks/.m2/repository/org/apache/maven/plugins/maven-gpg-plugin/3.0.1/maven-gpg-plugin-3.0.1.jar
08:57:53 [ERROR] urls[10] = file:/D:/Users/merks/.m2/repository/org/apache/maven/shared/maven-shared-utils/3.3.4/maven-shared-utils-3.3.4.jar
08:57:53 [ERROR] urls[11] = file:/D:/Users/merks/.m2/repository/org/codehaus/plexus/plexus-component-annotations/2.1.1/plexus-component-annotations-2.1.1.jar
08:57:53 [ERROR] urls[12] = file:/D:/Users/merks/.m2/repository/org/apache/maven/shared/maven-artifact-transfer/0.12.0/maven-artifact-transfer-0.12.0.jar
08:57:53 [ERROR] urls[13] = file:/D:/Users/merks/.m2/repository/org/apache/maven/shared/maven-common-artifact-filters/3.0.1/maven-common-artifact-filters-3.0.1.jar
08:57:53 [ERROR] urls[14] = file:/D:/Users/merks/.m2/repository/org/sonatype/sisu/sisu-inject-bean/1.4.2/sisu-inject-bean-1.4.2.jar
08:57:53 [ERROR] urls[15] = file:/D:/Users/merks/.m2/repository/org/sonatype/sisu/sisu-guice/2.1.7/sisu-guice-2.1.7-noaop.jar
08:57:53 [ERROR] urls[16] = file:/D:/Users/merks/.m2/repository/org/sonatype/plexus/plexus-sec-dispatcher/1.4/plexus-sec-dispatcher-1.4.jar
08:57:53 [ERROR] urls[17] = file:/D:/Users/merks/.m2/repository/org/sonatype/plexus/plexus-cipher/1.4/plexus-cipher-1.4.jar
08:57:53 [ERROR] urls[18] = file:/D:/Users/merks/.m2/repository/org/eclipse/platform/org.eclipse.equinox.p2.artifact.repository/1.4.500/org.eclipse.equinox.p2.artifact.repository-1.4.500.jar
08:57:53 [ERROR] urls[19] = file:/D:/Users/merks/.m2/repository/org/eclipse/platform/org.eclipse.equinox.common/3.16.200/org.eclipse.equinox.common-3.16.200.jar
08:57:53 [ERROR] urls[20] = file:/D:/Users/merks/.m2/repository/org/eclipse/platform/org.eclipse.equinox.registry/3.11.200/org.eclipse.equinox.registry-3.11.200.jar
08:57:53 [ERROR] urls[21] = file:/D:/Users/merks/.m2/repository/org/eclipse/platform/org.eclipse.equinox.p2.metadata/2.6.300/org.eclipse.equinox.p2.metadata-2.6.300.jar
08:57:53 [ERROR] urls[22] = file:/D:/Users/merks/.m2/repository/org/eclipse/platform/org.eclipse.equinox.p2.repository/2.6.200/org.eclipse.equinox.p2.repository-2.6.200.jar
08:57:53 [ERROR] urls[23] = file:/D:/Users/merks/.m2/repository/org/bouncycastle/bcprov-jdk18on/1.72/bcprov-jdk18on-1.72.jar
08:57:53 [ERROR] urls[24] = file:/D:/Users/merks/.m2/repository/org/bouncycastle/bcpg-jdk18on/1.72.2/bcpg-jdk18on-1.72.2.jar
08:57:53 [ERROR] urls[25] = file:/D:/Users/merks/.m2/repository/org/eclipse/platform/org.eclipse.core.jobs/3.13.100/org.eclipse.core.jobs-3.13.100.jar
08:57:53 [ERROR] urls[26] = file:/D:/Users/merks/.m2/repository/org/eclipse/platform/org.eclipse.equinox.preferences/3.10.100/org.eclipse.equinox.preferences-3.10.100.jar
08:57:53 [ERROR] urls[27] = file:/D:/Users/merks/.m2/repository/org/eclipse/platform/org.eclipse.equinox.p2.core/2.9.200/org.eclipse.equinox.p2.core-2.9.200.jar
08:57:53 [ERROR] urls[28] = file:/D:/Users/merks/.m2/repository/org/eclipse/platform/org.eclipse.equinox.security/1.3.1000/org.eclipse.equinox.security-1.3.1000.jar
08:57:53 [ERROR] urls[29] = file:/D:/Users/merks/.m2/repository/org/eclipse/platform/org.eclipse.osgi/3.18.100/org.eclipse.osgi-3.18.100.jar
08:57:53 [ERROR] urls[30] = file:/D:/Users/merks/.m2/repository/org/osgi/org.osgi.service.prefs/1.1.2/org.osgi.service.prefs-1.1.2.jar
08:57:53 [ERROR] urls[31] = file:/D:/Users/merks/.m2/repository/org/osgi/osgi.annotation/8.0.1/osgi.annotation-8.0.1.jar
08:57:53 [ERROR] urls[32] = file:/D:/Users/merks/.m2/repository/org/apache/maven/resolver/maven-resolver-util/1.9.1/maven-resolver-util-1.9.1.jar
08:57:53 [ERROR] urls[33] = file:/D:/Users/merks/.m2/repository/org/codehaus/plexus/plexus-interpolation/1.26/plexus-interpolation-1.26.jar
08:57:53 [ERROR] urls[34] = file:/D:/Users/merks/.m2/repository/org/codehaus/plexus/plexus-cipher/2.0/plexus-cipher-2.0.jar
08:57:53 [ERROR] urls[35] = file:/D:/Users/merks/.m2/repository/org/eclipse/tycho/sisu-equinox-launching/4.0.0-SNAPSHOT/sisu-equinox-launching-4.0.0-SNAPSHOT.jar
08:57:53 [ERROR] urls[36] = file:/D:/Users/merks/.m2/repository/org/apache/commons/commons-exec/1.3/commons-exec-1.3.jar
08:57:53 [ERROR] urls[37] = file:/D:/Users/merks/.m2/repository/org/eclipse/tycho/sisu-osgi-api/4.0.0-SNAPSHOT/sisu-osgi-api-4.0.0-SNAPSHOT.jar
08:57:53 [ERROR] urls[38] = file:/D:/Users/merks/.m2/repository/org/eclipse/tycho/tycho-targetplatform/4.0.0-SNAPSHOT/tycho-targetplatform-4.0.0-SNAPSHOT.jar
08:57:53 [ERROR] urls[39] = file:/D:/Users/merks/.m2/repository/org/eclipse/platform/org.eclipse.equinox.frameworkadmin/2.2.100/org.eclipse.equinox.frameworkadmin-2.2.100.jar
08:57:53 [ERROR] urls[40] = file:/D:/Users/merks/.m2/repository/org/eclipse/platform/org.eclipse.equinox.p2.metadata.repository/1.4.100/org.eclipse.equinox.p2.metadata.repository-1.4.100.jar
08:57:53 [ERROR] urls[41] = file:/D:/Users/merks/.m2/repository/org/eclipse/platform/org.eclipse.equinox.p2.jarprocessor/1.2.300/org.eclipse.equinox.p2.jarprocessor-1.2.300.jar
08:57:53 [ERROR] urls[42] = file:/D:/Users/merks/.m2/repository/org/eclipse/platform/org.eclipse.equinox.p2.publisher/1.7.200/org.eclipse.equinox.p2.publisher-1.7.200.jar
08:57:53 [ERROR] urls[43] = file:/D:/Users/merks/.m2/repository/org/eclipse/platform/org.eclipse.equinox.p2.director/2.5.400/org.eclipse.equinox.p2.director-2.5.400.jar
08:57:53 [ERROR] urls[44] = file:/D:/Users/merks/.m2/repository/org/ow2/sat4j/org.ow2.sat4j.core/2.3.6/org.ow2.sat4j.core-2.3.6.jar
08:57:53 [ERROR] urls[45] = file:/D:/Users/merks/.m2/repository/org/ow2/sat4j/org.ow2.sat4j.pb/2.3.6/org.ow2.sat4j.pb-2.3.6.jar
08:57:53 [ERROR] urls[46] = file:/D:/Users/merks/.m2/repository/org/eclipse/platform/org.eclipse.equinox.p2.updatesite/1.2.300/org.eclipse.equinox.p2.updatesite-1.2.300.jar
08:57:53 [ERROR] urls[47] = file:/D:/Users/merks/.m2/repository/org/eclipse/platform/org.eclipse.equinox.p2.publisher.eclipse/1.4.100/org.eclipse.equinox.p2.publisher.eclipse-1.4.100.jar
08:57:53 [ERROR] urls[48] = file:/D:/Users/merks/.m2/repository/org/eclipse/platform/org.eclipse.equinox.frameworkadmin.equinox/1.2.200/org.eclipse.equinox.frameworkadmin.equinox-1.2.200.jar
08:57:53 [ERROR] urls[49] = file:/D:/Users/merks/.m2/repository/org/eclipse/platform/org.eclipse.equinox.simpleconfigurator.manipulator/2.2.0/org.eclipse.equinox.simpleconfigurator.manipulator-2.2.0.jar
08:57:53 [ERROR] urls[50] = file:/D:/Users/merks/.m2/repository/org/osgi/org.osgi.service.component/1.5.0/org.osgi.service.component-1.5.0.jar
08:57:53 [ERROR] urls[51] = file:/D:/Users/merks/.m2/repository/org/osgi/org.osgi.namespace.extender/1.0.1/org.osgi.namespace.extender-1.0.1.jar
08:57:53 [ERROR] urls[52] = file:/D:/Users/merks/.m2/repository/org/eclipse/platform/org.eclipse.equinox.concurrent/1.2.100/org.eclipse.equinox.concurrent-1.2.100.jar
08:57:53 [ERROR] urls[53] = file:/D:/Users/merks/.m2/repository/org/eclipse/tycho/p2-maven-plugin/4.0.0-SNAPSHOT/p2-maven-plugin-4.0.0-SNAPSHOT.jar
08:57:53 [ERROR] urls[54] = file:/D:/Users/merks/.m2/repository/org/eclipse/platform/org.eclipse.equinox.p2.engine/2.7.500/org.eclipse.equinox.p2.engine-2.7.500.jar
08:57:53 [ERROR] urls[55] = file:/D:/Users/merks/.m2/repository/org/eclipse/platform/org.eclipse.osgi.compatibility.state/1.2.700/org.eclipse.osgi.compatibility.state-1.2.700.jar
08:57:53 [ERROR] urls[56] = file:/D:/Users/merks/.m2/repository/org/eclipse/tycho/sisu-osgi-connect/4.0.0-SNAPSHOT/sisu-osgi-connect-4.0.0-SNAPSHOT.jar
08:57:53 [ERROR] urls[57] = file:/D:/Users/merks/.m2/repository/org/eclipse/platform/org.eclipse.equinox.p2.touchpoint.natives/1.4.400/org.eclipse.equinox.p2.touchpoint.natives-1.4.400.jar
08:57:53 [ERROR] urls[58] = file:/D:/Users/merks/.m2/repository/org/eclipse/platform/org.eclipse.equinox.p2.touchpoint.eclipse/2.3.300/org.eclipse.equinox.p2.touchpoint.eclipse-2.3.300.jar
08:57:53 [ERROR] urls[59] = file:/D:/Users/merks/.m2/repository/org/eclipse/platform/org.eclipse.equinox.p2.garbagecollector/1.2.0/org.eclipse.equinox.p2.garbagecollector-1.2.0.jar
08:57:53 [ERROR] urls[60] = file:/D:/Users/merks/.m2/repository/org/eclipse/platform/org.eclipse.equinox.simpleconfigurator/1.4.100/org.eclipse.equinox.simpleconfigurator-1.4.100.jar
08:57:53 [ERROR] urls[61] = file:/D:/Users/merks/.m2/repository/org/eclipse/platform/org.eclipse.equinox.app/1.6.200/org.eclipse.equinox.app-1.6.200.jar
08:57:53 [ERROR] urls[62] = file:/D:/Users/merks/.m2/repository/org/apache/felix/org.apache.felix.scr/2.2.4/org.apache.felix.scr-2.2.4.jar
08:57:53 [ERROR] urls[63] = file:/D:/Users/merks/.m2/repository/org/codehaus/mojo/animal-sniffer-annotations/1.9/animal-sniffer-annotations-1.9.jar
08:57:53 [ERROR] urls[64] = file:/D:/Users/merks/.m2/repository/org/bouncycastle/bcpg-jdk15on/1.70/bcpg-jdk15on-1.70.jar
08:57:53 [ERROR] urls[65] = file:/D:/Users/merks/.m2/repository/org/bouncycastle/bcprov-jdk15on/1.70/bcprov-jdk15on-1.70.jar
08:57:53 [ERROR] urls[66] = file:/D:/Users/merks/.m2/repository/org/eclipse/platform/org.eclipse.core.runtime/3.26.0/org.eclipse.core.runtime-3.26.0.jar
08:57:53 [ERROR] urls[67] = file:/D:/Users/merks/.m2/repository/org/eclipse/platform/org.eclipse.core.contenttype/3.8.200/org.eclipse.core.contenttype-3.8.200.jar
08:57:53 [ERROR] urls[68] = file:/D:/Users/merks/.m2/repository/biz/aQute/bnd/biz.aQute.bndlib/6.3.1/biz.aQute.bndlib-6.3.1.jar
08:57:53 [ERROR] urls[69] = file:/D:/Users/merks/.m2/repository/biz/aQute/bnd/biz.aQute.bnd.util/6.3.1/biz.aQute.bnd.util-6.3.1.jar
08:57:53 [ERROR] urls[70] = file:/D:/Users/merks/.m2/repository/org/osgi/org.osgi.service.repository/1.1.0/org.osgi.service.repository-1.1.0.jar
08:57:53 [ERROR] urls[71] = file:/D:/Users/merks/.m2/repository/org/osgi/org.osgi.util.function/1.2.0/org.osgi.util.function-1.2.0.jar
08:57:53 [ERROR] urls[72] = file:/D:/Users/merks/.m2/repository/org/osgi/org.osgi.util.promise/1.2.0/org.osgi.util.promise-1.2.0.jar
08:57:53 [ERROR] urls[73] = file:/D:/Users/merks/.m2/repository/org/eclipse/platform/org.eclipse.equinox.p2.director.app/1.2.200/org.eclipse.equinox.p2.director.app-1.2.200.jar
08:57:53 [ERROR] urls[74] = file:/D:/Users/merks/.m2/repository/org/eclipse/platform/org.eclipse.equinox.p2.repository.tools/2.3.100/org.eclipse.equinox.p2.repository.tools-2.3.100.jar
08:57:53 [ERROR] urls[75] = file:/D:/Users/merks/.m2/repository/commons-codec/commons-codec/1.15/commons-codec-1.15.jar
08:57:53 [ERROR] Number of foreign imports: 1
08:57:53 [ERROR] import: Entry[import  from realm ClassRealm[project>org.eclipse.tycho.maven:org.eclipse.tycho.maven.parent:1.0.0-SNAPSHOT, parent: ClassRealm[maven.api, parent: null]]]
08:57:53 [ERROR] 
```
I've seen various forms of this problem over the last week.

Reply:
Can you try adding `org.eclipse.equinox.internal.p2.artifact.repository.simple` to the exported packages of https://github.com/eclipse-tycho/tycho/blob/master/tycho-maven-plugin/src/main/resources/META-INF/maven/extension.xml ?

Reply:
You're awesome! 🥇